accepted/2025/PrunePackageReference-with-direct-PackageReference.md

```diff
@@ -0,0 +1,248 @@
+| dotnet list package | Show | Similar as Package Manager UI, installed tab. The list package has a requested/resolved column. The resolved column should match the requested, as well as an additional indicator similar to auto-referenced. |
+| dotnet list package --outdated | Show | A customer needs to be able to update, as an update may make a package not pruned anymore. The resolved column should match the requested, as well as an additional indicator similar to auto-referenced. |
+| dotnet nuget why | Show | Similar reasoning as list package. |
+| Solution Explorer | Show | The package needs to indicate why it was not "resolved". The solution explorer currently adds a warning icon next to unresolved packages (ie, not found in the project.assets.json file). We should change the indicator to something different. |
+| GetInstalledPackagesAsync | Show | Depends on the scenario, they may be trying to achieve similar behavior to what the PM UI and list package commands provide. The [NuGetInstalledPackage](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.VisualStudio.Contracts/NuGetInstalledPackage.cs) type from the INuGetProjectService (that test explorer and roslyn use), will have the following changes: `Version` => null, the documentation will be updated to say the package is in either an error case or pruned, InstallPath => null, similarly if the package does not exist, the installation path does not exist. |
+
+Given that pruning is an offline operation, ie NuGet does not talk to the sources for pruned packages, we never validate that the requested package exists and the specified version may be incorrect. This is likely to not be of too much consequence, since it's only really possible with hand-editing, and the impact of it is minimal since the package is not used anyways.
+
+## Prior Art
+
+N/A
+
+## Unresolved Questions
+
+- Should we prune direct packages? If the answer is yes, the following unresolved questions apply:
+  - What should `dotnet list package` show for direct pruned packages? Should it show the version specified as requested?
+    - It can be marked in a similar way that auto-referenced packages are marked. `> Newtonsoft.Json           (A)   [13.0.2, )   13.0.2`
+  - What should the GetInstalledPackages return for the pruned direct packages?
+  - Should we add a setting to prune direct package references?
+  - Validate how 3rd party components would behave with unresolved direct package references.
+
+## Future Possibilities
+
+<!-- What future possibilities can you think of that this proposal would help with? -->
+
+- If we prune direct packages:
+  - Visualize pruned packages in the PM UI Installed/Updates tabs. Note that the PM UI does not have visualization support for multi-targeting, so the heuristic here should be similar to the one for the NU1510 warning.
```

Comment:
What would the visualization support look like? I can imagine the Installed Packages List could show the pruned packages in some way that indicates they're not applicable due to pruning (perhaps a different group for those, or just a different background and some indicator).
For Package Update, perhaps a prompt or acknowledgement pop-up when a pruned package is being promoted? 
Other ideas? We can discuss later once your improvements are shipped as well.

Reply:
> I can imagine the Installed Packages List could show the pruned packages in some way that indicates they're not applicable due to pruning

Yeah, probably something like that. 
I think the urgency of the future work is also different depending on whether we prune completely, orf if we apply IncludeAssets and PrivateAssets (more in the former one, less critical in the latter one).

Reply:
> For Package Update, perhaps a prompt or acknowledgement pop-up when a pruned package is being promoted?

The update one is tricky, you could be updating a package out of the range, so we'd need to mindful there. 
I don't really want prompts tbh. 
I feel like I'd rather fail the operation completely since it wouldn't have an effect. 

>  We can discuss later once your improvements are shipped as well.
Yes for sure. This is all assuming any of the direct pruning changes are approved.